core/src/Streamly/Internal/Data/Unbox/TH.hs

```diff
@@ -0,0 +1,412 @@
+        [] -> [|pure ()|]
+        _ ->
+            letE
+                (mkOffsetDecls tagSize fields)
+                (doE $ map (noBindS . pokeField) [0 .. (numFields - 1)])
+
+    where
+
+    numFields = length fields
+    pokeField i =
+        [|pokeByteIndex
+              $(varE (mkOffsetName i))
+              $(varE _arr)
+              $(varE (mkFieldName i))|]
+
+mkPokeMatch :: Name -> Int -> Q Exp -> Q Match
+mkPokeMatch cname numFields exp0 =
+    match
+        (conP cname (map varP (map mkFieldName [0 .. (numFields - 1)])))
+        (normalB exp0)
+        []
+
+mkPokeExpr :: Type -> [DataCon] -> Q Exp
+mkPokeExpr headTy cons =
+    case cons of
+        [] ->
+            [|error
+                  ("Attempting to poke type with no constructors (" ++
+                   $(lift (pprint headTy)) ++ ")")|]
+        [(DataCon _ _ _ [])] -> [|pure ()|]
+        [(DataCon cname _ _ fields)] ->
+            caseE
+                (varE _val)
+                [mkPokeMatch cname (length fields) (mkPokeExprFields 0 fields)]
+        _ ->
+            caseE
+                (varE _val)
+                (map (\(tagVal, (DataCon cname _ _ fields)) ->
+                          mkPokeMatch
+                              cname
+                              (length fields)
+                              (doE [ noBindS $ mkPokeExprTag tagType tagVal
+                                   , noBindS $ mkPokeExprFields tagSize fields
+                                   ]))
+                     (zip [0 ..] cons))
+
+    where
+
+    lenCons = length cons
+    tagType = getTagType lenCons
+    tagSize = getTagSize lenCons
+
+
+--------------------------------------------------------------------------------
+-- Main
+--------------------------------------------------------------------------------
+
+-- XXX Document this function properly
+makeUnboxWith :: Cxt -> Type -> [DataCon] -> Q [Dec]
+makeUnboxWith preds headTy cons = do
+    sizeOfMethod <- mkSizeOfExpr headTy cons
+    peekMethod <- mkPeekExpr headTy cons
+    pokeMethod <- mkPokeExpr headTy cons
+    let methods =
+            [ FunD 'sizeOf [Clause [WildP] (NormalB sizeOfMethod) []]
+            , FunD
+                  'peekByteIndex
+                  [ Clause
+                        (if isUnitType cons
+                             then [WildP, WildP]
+                             else [VarP _initialOffset, VarP _arr])
+                        (NormalB peekMethod)
+                        []
+                  ]
+            , FunD
+                  'pokeByteIndex
+                  [ Clause
+                        (if isUnitType cons
+                             then [WildP, WildP, WildP]
+                             else [VarP _initialOffset, VarP _arr, VarP _val])
+                        (NormalB pokeMethod)
+                        []
+                  ]
+            ]
+    return [plainInstanceD preds (AppT (ConT ''Unbox) headTy) methods]
+
+makeUnbox :: Name -> Q [Dec]
```

Comment:
1. Name it `deriveStore` instead.
2. Write documentation on how to use it. Does it require the type variables of the data type to be Unbox instances?
3. We need to export it from Array modules so that users can find it and use it.
4. Write guidelines in Unbox doc that TH should be preferred as it is faster